src/include/kdbprivate.h

```diff
@@ -210,6 +220,23 @@ struct _Key
 	 */
 	size_t dataSize;
 
+	/**
+	 * Reference counter
+	 */
+	uint16_t refs;
+
+	uint16_t flags;
```

Comment:
We can specify here how many bits we actually need.

Reply:
What do you mean? We already specify that it's 16 bits?

Reply:
We only need 3 bits or so.

```suggestion
	uint16_t flags : 3;
```

Reply:
AFAICT the bitfields in C are a total mess. From what I found online, the actual memory layout of bit fields is entirely implementation defined. So adding the `: 3` may make the struct incompatible between different compilers in theory.

Reply:
Okay but if we go this route, I'd rather split up each flag as a bitfield of 1.

```suggestion
	bool isInMmap : 1;
	bool anotherFlag : 1;
	bool yetAnotherFlag : 1;
```

Would improved readability of the code IMHO.

Reply:
Great suggestion. Doesn't need to be within this PR, unless you touch this anyway.

Reply:
> Doesn't need to be within this PR, unless you touch this anyway.

Not much of a problem for me, they have already been touched so I can touch them again, at least for the COW structures. For `Key` and `KeySet` I need to look how many tests depend on `flags` field being there.

> may make the struct incompatible between different compilers in theory

Yes, but is this really a concern? We're talking about purely internal datastructures here. Is it likely, that parts of Elektra will be built using different compilers?

Reply:
I looked into it a bit more and while the C standard seems to have only very vague definitions of bitfields, the AMD64 ABI is more specific. So I think we should do

```suggestion
	bool isInMmap : 1;
	int : 15;
```

The unnamed `int : 15` field reserves 15 more bits for future use. We don't need to think about the type and can just use a simple `int`, because the AMD64 ABI says: "Unnamed bit-fields’ types do not affect the alignment of a structure or union."

Adding more flags should be a completely transparent change (even for `mmap`) by just adding another `bool` and reduce the size of the unnamed field.

I also found the following infos, which all point toward `bool flag : 1` being better than `uint16_t flags : 3` in general:
- The C standard only explicitly allows bitfields with types `bool`, `signed int` and `unsigned int`. For just `int` it is implementation defined, whether the field is signed or unsigned. Using any other types would be entirely implementation defined.
- The memory layout is actually not entirely implementation defined. The C standard says that adjacent bit-fields in a struct should be packed together. But it is implementation defined, whether a bit-field can overlap two storage units. So by only using bit-fields of size 1, we are definitely covered by the standard and everything must be packed in the minimum possible space.

Reply:
> Yes, but is this really a concern? We're talking about purely internal datastructures here. Is it likely, that parts of Elektra will be built using different compilers?

See my previous comment. Using `bool` should be fine on every compiler.

I wasn't super concerned about the issue anyway. I just wanted to mention that bit-fields are not fully specified in the standard, since normally try to avoid implementation defined behaviour.

Reply:
Done.